Implement ElementInternals

shvaikalesh opened the pull request:
#### c51d445d35e1f4407bd8fe107ebe158363e7ae8f
<pre>
Implement ElementInternals
<a href="https://bugs.webkit.org/show_bug.cgi?id=197960">https://bugs.webkit.org/show_bug.cgi?id=197960</a>

Reviewed by Ryosuke Niwa.

This change implements ElementInternals interface [1] with a `shadowRoot` getter only,
as well as allowing to disable attachInternals() via `disabledFeatures` property.

To avoid breaking web sites and polyfills that rely on form-associated custom elements
being fully supported if attachInternals() is present, ElementInternals is carefully
hidden behind a runtime flag.

To ensure attachInternals() can be called in a custom element constructor but not if
upgrading fails, the spec introduced a fifth custom element state of &quot;precustomized&quot;,
which is very similar to &quot;failed&quot;.

This change avoids adding CustomElementState::Precustomized to preserve free bits in
Node::RareDataBitFields, and instead leverages that &quot;precustomized&quot; state is set
only in upgrade algorithm, is very similar to &quot;failed&quot;, and we can distinguish them
by checking Node::IsUnknownElement flag.

For the same reason, not to increase sizeof(ElementRareData), ElementInternals attached
flag is defined on CustomElementReactionQueue, which led to loosening up assertions in
Element::reactionQueue() as it now may be invoked on a built-in element.

Also, the introduction of CustomElementState::PrecustomizedOrFailed led to adjusting
custom element state checks, bringing us closer to the spec steps / terminology by
removing inverted checks in upgradeElement() / at isUndefinedCustomElement() call sites.

[1] <a href="https://html.spec.whatwg.org/multipage/custom-elements.html#the-elementinternals-interface">https://html.spec.whatwg.org/multipage/custom-elements.html#the-elementinternals-interface</a>

Tests: imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry.html
       imported/w3c/web-platform-tests/custom-elements/HTMLElement-attachInternals.html
       imported/w3c/web-platform-tests/custom-elements/element-internals-shadowroot.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/HTMLElement-attachInternals-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/element-internals-shadowroot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-NotSupportedError-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-form-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-labels-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-setFormValue-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-validation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/form-associated-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/form-disabled-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state.tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state.tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::JSCustomElementInterface):
(WebCore::JSCustomElementInterface::upgradeElement):
(WebCore::JSCustomElementInterface::setFormAssociatedCallback):
(WebCore::JSCustomElementInterface::setFormResetCallback):
(WebCore::JSCustomElementInterface::setFormDisabledCallback):
(WebCore::JSCustomElementInterface::setFormStateRestoreCallback):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
(WebCore::JSCustomElementInterface::disableElementInternals):
(WebCore::JSCustomElementInterface::isElementInternalsDisabled const):
(WebCore::JSCustomElementInterface::setIsFormAssociated):
* Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
(WebCore::JSCustomElementRegistry::define):
* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp: Added.
(WebCore::JSElementInternals::visitAdditionalChildren):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::isDefinedElement):
* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementReactionQueue::isElementInternalsDisabled const):
(WebCore::CustomElementReactionQueue::isElementInternalsAttached const):
(WebCore::CustomElementReactionQueue::setElementInternalsAttached):
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::document const):
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setIsFailedCustomElement):
(WebCore::Element::setIsFailedOrPrecustomizedCustomElementWithoutClearingReactionQueue):
(WebCore::Element::clearReactionQueueFromFailedOrPrecustomizedCustomElement):
(WebCore::Element::enqueueToUpgrade):
(WebCore::Element::reactionQueue const):
(WebCore::Element::setIsFailedCustomElementWithoutClearingReactionQueue): Deleted.
(WebCore::Element::clearReactionQueueFromFailedCustomElement): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp: Added.
(WebCore::ElementInternals::shadowRoot const):
* Source/WebCore/dom/ElementInternals.h: Added.
* Source/WebCore/dom/ElementInternals.idl: Added.
* Source/WebCore/dom/Node.h:
(WebCore::Node::isUncustomizedCustomElement const):
(WebCore::Node::isFailedCustomElement const):
(WebCore::Node::isFailedOrPrecustomizedCustomElement const):
(WebCore::Node::isPrecustomizedCustomElement const):
(WebCore::Node::isPrecustomizedOrDefinedCustomElement const):
(WebCore::Node::isUndefinedCustomElement const): Deleted.
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::attachInternals):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::customElementState):

Canonical link: <a href="https://commits.webkit.org/253303@main">https://commits.webkit.org/253303@main</a>
</pre>

rniwa commented on Source/WebCore/dom/ElementInternals.cpp:
```diff
@@ -0,0 +1,47 @@
+
+ShadowRoot* ElementInternals::shadowRoot() const
+{
+    if (auto* element = m_element.get()) {
```
Can we instead use early returns? Also, please use RefPtr here.
i.e.
```cpp
RefPtr element = m_element.get();
if (!element)
    return nullptr;
auto* shadowRoot = element->shadowRoot();
if (!shadowRoot || !shadowRoot->isAvailableToElementInternals)
    return nullptr;
return shadowRoot;
```

rniwa commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1124,6 +1126,25 @@ bool HTMLElement::shouldExtendSelectionToTargetNode(const Node& targetNode, cons
+{
+    auto* queue = reactionQueue();
+    if (!queue)
+        return Exception { NotSupportedError };
```
Can we give it a useful error message? e.g. "attachInternals is only supported on a custom element instance"

rniwa commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1124,6 +1126,25 @@ bool HTMLElement::shouldExtendSelectionToTargetNode(const Node& targetNode, cons
+        return Exception { NotSupportedError };
+
+    if (queue->isElementInternalsDisabled())
+        return Exception { NotSupportedError };
```
"attachInternals is disabled"

rniwa commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1124,6 +1126,25 @@ bool HTMLElement::shouldExtendSelectionToTargetNode(const Node& targetNode, cons
+        return Exception { NotSupportedError };
+
+    if (queue->isElementInternalsAttached())
+        return Exception { NotSupportedError };
```
"There is already an existing ElementInternals"

rniwa commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1124,6 +1126,25 @@ bool HTMLElement::shouldExtendSelectionToTargetNode(const Node& targetNode, cons
+        return Exception { NotSupportedError };
+
+    if (!isPrecustomizedOrDefinedCustomElement())
+        return Exception { NotSupportedError };
```
"Custom element is in an invalid state"

rniwa commented on Source/WebCore/dom/ElementInternals.idl:
```diff
@@ -0,0 +1,32 @@
+[
+    GenerateIsReachable=ImplElementRoot,
+    Exposed=Window,
+    JSCustomMarkFunction,
```
Can we use `GenerateAddOpaqueRoot` instead of `JSCustomMarkFunction`?

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2764,21 +2766,22 @@ void Element::setIsDefinedCustomElement(JSCustomElementInterface& elementInterfa
 }
 
-void Element::clearReactionQueueFromFailedCustomElement()
+void Element::clearReactionQueueFromFailedOrPrecustomizedCustomElement()
```
I don't think we should rename this function.
Whenever we invoke this function, we know "this" is a failed custom element.

rniwa reviewed (approved):
r=me

shvaikalesh committed:
Implement ElementInternals
https://bugs.webkit.org/show_bug.cgi?id=197960

Reviewed by Ryosuke Niwa.

This change implements ElementInternals interface [1] with a `shadowRoot` getter only,
as well as allowing to disable attachInternals() via `disabledFeatures` property.

To avoid breaking web sites and polyfills that rely on form-associated custom elements
being fully supported if attachInternals() is present, ElementInternals is carefully
hidden behind a runtime flag.

To ensure attachInternals() can be called in a custom element constructor but not if
upgrading fails, the spec introduced a fifth custom element state of "precustomized",
which is very similar to "failed".

This change avoids adding CustomElementState::Precustomized to preserve free bits in
Node::RareDataBitFields, and instead leverages that "precustomized" state is set
only in upgrade algorithm, is very similar to "failed", and we can distinguish them
by checking Node::IsUnknownElement flag.

For the same reason, not to increase sizeof(ElementRareData), ElementInternals attached
flag is defined on CustomElementReactionQueue, which led to loosening up assertions in
Element::reactionQueue() as it now may be invoked on a built-in element.

Also, the introduction of CustomElementState::PrecustomizedOrFailed led to adjusting
custom element state checks, bringing us closer to the spec steps / terminology by
removing inverted checks in upgradeElement() / at isUndefinedCustomElement() call sites.

[1] https://html.spec.whatwg.org/multipage/custom-elements.html#the-elementinternals-interface

Tests: imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry.html
       imported/w3c/web-platform-tests/custom-elements/HTMLElement-attachInternals.html
       imported/w3c/web-platform-tests/custom-elements/element-internals-shadowroot.html

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/HTMLElement-attachInternals-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/element-internals-shadowroot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-NotSupportedError-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-form-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-labels-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-setFormValue-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-validation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/form-associated-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/form-disabled-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state.tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state.tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::JSCustomElementInterface):
(WebCore::JSCustomElementInterface::upgradeElement):
(WebCore::JSCustomElementInterface::setFormAssociatedCallback):
(WebCore::JSCustomElementInterface::setFormResetCallback):
(WebCore::JSCustomElementInterface::setFormDisabledCallback):
(WebCore::JSCustomElementInterface::setFormStateRestoreCallback):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
(WebCore::JSCustomElementInterface::disableElementInternals):
(WebCore::JSCustomElementInterface::isElementInternalsDisabled const):
(WebCore::JSCustomElementInterface::setIsFormAssociated):
* Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
(WebCore::JSCustomElementRegistry::define):
* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp: Added.
(WebCore::JSElementInternals::visitAdditionalChildren):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::isDefinedElement):
* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementReactionQueue::isElementInternalsDisabled const):
(WebCore::CustomElementReactionQueue::isElementInternalsAttached const):
(WebCore::CustomElementReactionQueue::setElementInternalsAttached):
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::document const):
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setIsFailedCustomElement):
(WebCore::Element::setIsFailedOrPrecustomizedCustomElementWithoutClearingReactionQueue):
(WebCore::Element::clearReactionQueueFromFailedOrPrecustomizedCustomElement):
(WebCore::Element::enqueueToUpgrade):
(WebCore::Element::reactionQueue const):
(WebCore::Element::setIsFailedCustomElementWithoutClearingReactionQueue): Deleted.
(WebCore::Element::clearReactionQueueFromFailedCustomElement): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp: Added.
(WebCore::ElementInternals::shadowRoot const):
* Source/WebCore/dom/ElementInternals.h: Added.
* Source/WebCore/dom/ElementInternals.idl: Added.
* Source/WebCore/dom/Node.h:
(WebCore::Node::isUncustomizedCustomElement const):
(WebCore::Node::isFailedCustomElement const):
(WebCore::Node::isFailedOrPrecustomizedCustomElement const):
(WebCore::Node::isPrecustomizedCustomElement const):
(WebCore::Node::isPrecustomizedOrDefinedCustomElement const):
(WebCore::Node::isUndefinedCustomElement const): Deleted.
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::attachInternals):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::customElementState):

Canonical link: https://commits.webkit.org/253303@main
